Sources/NIOCore/ByteBuffer-QUICLengthPrefix.swift

```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    /// Write a QUIC variable-length integer prefixed buffer.
+    ///
+    /// [RFC 9000 § 16](https://www.rfc-editor.org/rfc/rfc9000.html#section-16)
+    ///
+    /// Write `buffer` prefixed with the length of buffer as a QUIC variable-length integer
+    /// - Parameter buffer: The buffer to be written
+    /// - Returns: The total bytes written. This is the bytes needed to write the length, plus the length of the buffer itself
+    @discardableResult
+    public mutating func writeQUICLengthPrefixed(_ buffer: ByteBuffer) -> Int {
+        var written = 0
+        written += self.writeQUICVariableLengthInteger(buffer.readableBytes)
+        written += self.writeImmutableBuffer(buffer)
+        return written
+    }
+
+    /// Prefixes a message written by `writeMessage` with the number of bytes written as a QUIC variable-length integer
+    /// - Parameters:
+    ///     - writeMessage: A closure that takes a buffer, writes a message to it and returns the number of bytes written
+    /// - Returns: Number of total bytes written
+    /// - Note: Because the length of the message is not known upfront, 4 bytes will be used for encoding the length even if that may not be necessary. If you know the length of your message, prefer ``writeQUICLengthPrefixed(_:)`` instead
+    @discardableResult
+    @inlinable
+    public mutating func writeQUICLengthPrefixed(
+        writeMessage: (inout ByteBuffer) throws -> Int
+    ) rethrows -> Int {
+        var totalBytesWritten = 0
+
+        let lengthPrefixIndex = self.writerIndex
+        // Write 4 bytes as a placeholder
+        // This is enough for upto 1gb of data
+        // It is very unlikely someone is trying to write more than that, if they are then we will catch it later and do a memmove
+        // Reserving 8 bytes now would require either wasting bytes for the majority of use cases, or doing a memmove for a majority of usecases
+        // This way the majority only waste 0-3 bytes, and never incur a memmove, and the minority get a memmove
+        totalBytesWritten += self.writeBytes([0, 0, 0, 0])
+
+        let startWriterIndex = self.writerIndex
+        let messageLength = try writeMessage(&self)
+        let endWriterIndex = self.writerIndex
+
+        totalBytesWritten += messageLength
+
+        let actualBytesWritten = endWriterIndex - startWriterIndex
+        assert(
+            actualBytesWritten == messageLength,
+            "writeMessage returned \(messageLength) bytes, but actually \(actualBytesWritten) bytes were written, but they should be the same"
+        )
+
+        if messageLength >= 1_073_741_823 {
+            // This message length cannot fit in the 4 bytes which we reserved. We need to make more space
+            // Reserve 4 more bytes
+            totalBytesWritten += self.writeBytes([0, 0, 0, 0])
+            // Move the message forward by 4 bytes
+            self.withUnsafeMutableReadableBytes { pointer in
+                _ = memmove(
```

Comment:
Rather than bury this in here, please factor this out into a `fileprivate` helper method.

Reply:
We could but it will need to be internal if we want to keep inlineable

Reply:
That's fine, leading underscore works great too.